### DISCUSSION TITLE
useControl relationship to field meta (value, valid, etc.)

### DISCUSSION THREAD

**Discussion by halljus** [edmundhung/conform#475] [Q&A]
I'm taking `useControl` for a test drive and noticed a couple things. First, when I change the control value with `control.change`, it doesn't change the field value. At first I thought this was fine, I'd just use the `control.value` everywhere I was previously using `field.value`. But then I noticed `field.valid` doesn't seem to stay in sync either. For example, if I try to submit an invalid field, `field.valid` will change from `true` to `false` as expected, but when I change to a valid value (and call `control.change`), `field.valid` remains `false` (probably due to the first issue where the `value` is not updated?). So now I'm just wondering about the relationship between the control and the field... 🤔

**Comment by edmundhung** [comment]
If you see `field.value` and `control.value` not in sync, it's likely something goes wrong with the register step. (e.g. register on the wrong element). `field.value` is always a sync of the FormData API which reads directly from the dom. While `control.value` is just a react state which is suppose to be synced with the input you registered when you call `control.change()`. Which UI library are you using? I think I should also start putting a PR up with all the examples updated to use the useControl hook.

**Reply by halljus** [reply]
In this particular case, we're using [react-bootstrap-typeahead](https://github.com/ericgio/react-bootstrap-typeahead/). Here's some pseudo-code:

```tsx
const thingsTypeaheadRef = React.useRef<TypeaheadRef>(null);

const [form, fields] = useForm({
  // other options...
  defaultValue: { thingIds: '' },
});

const thingIdsControl = useControl(fields.thingIds);

// Parsing the value for easier consumption
const selectedThingIdsArray = JSON.parse(thingIdsControl.value ?? '[]') as string[];

<Typeahead
  // some other options...
  ref={thingsTypeaheadRef}
  selected={
    selectedThingIdsArray
      .map((thingOptionId) => thingOptions.find((thingOption) => thingOption.value === thingOptionId))
      .filter(Boolean)
  }
  onChange={(selectedOptions) => {
    thingIdsControl.change(
      JSON.stringify(selectedOptions.map((option) => option.value)),
    );
  }}
  isInvalid={!fields.thingIds.valid}
  multiple
/>

<input
  className="visually-hidden"
  tabIndex={-1}
  name={fields.thingIds.name}
  defaultValue={fields.thingIds.initialValue}
  ref={thingIdsControl.register}
  value={thingIdsControl.value}
  onFocus={() => {
    thingsTypeaheadRef.current?.focus();
  }}
/>
```

**Reply by edmundhung** [reply]
This looks correct to me 🤔 Maybe you can try passing the `value` as `defaultValue` on the hidden input and see if the input value is updated correctly with `control.change()`.

If it is registered correctly, I expect the input value to be updated as you call `control.change()`.

```tsx
<input
  className="visually-hidden"
  tabIndex={-1}
  name={fields.thingIds.name}
  defaultValue={fields.thingIds.initialValue}
  ref={thingIdsControl.register}
  defaultValue={thingIdsControl.value}
  onFocus={() => {
    thingsTypeaheadRef.current?.focus();
  }}
/>
```

FYI, you can also model this as a multiple select element. Then there is no need to serialize it as one single string.

**Reply by halljus** [reply]
Changing `defaultValue` from `fields.thingIds.initialValue` to `thingIdsControl.value` didn't seem to have any effect.

And yeah, I still need to try modeling it as multi-select. I'll do that next, but I would think the string model should have worked.

**Reply by edmundhung** [reply]
I might need a repo to debug further. But please confirm if the input value is indeed updated. I think the underlying issue here should just be the input value not updating properly and hence `field.value` and `field.valid` looks out of sync. Otherwise this might not be related to `useControl`.

**Reply by halljus** [reply]
In my original code (above), the `control.value` (hidden input value) updates correctly, but the field meta value doesn't. When I change `defaultValue` from `fields.thingIds.initialValue` to `thingIdsControl.value`, it has no effect - but I had left `value={thingIdsControl.value}` in place. If I also remove `value` all together, they are in sync (and `form.valid`, too).

However, getting past that has revealed that the `thingIds` field is not remaining in `form.value` when conditionally rendered (bringing us full circle). 😅 I can make a repro in a bit.

**Reply by halljus** [reply]
~~This isn't quite _exactly_ what's happening in the real app, but I think it's close enough to illustrate the conditional rendering issue.~~

~~https://stackblitz.com/edit/remix-run-remix-dpreq3?file=package.json,app%2Froutes%2F_index.tsx~~

Actually here, this one matches the real app...

https://stackblitz.com/edit/remix-run-remix-kt8ajd?file=app%2Froutes%2F_index.tsx

**Reply by edmundhung** [reply]
I have to say this is working as expected and you are seeing the limitation of the current design 😅 

1. `form.value` or any field `value` is a sync of the FormData API. Its value totally reflect the DOM with one exception: First document load. In this case, we have no access to the dom and so the value you get is just what you passed to the `defaultValue` option.
2. You will also see the `toppingIds` slightly out of sync when you are selecting different option. This is what happened when you select an option: (1) Input event dispatched by the radio button (2) Conform read the DOM with the FormData API and update the form value (3) The form value is updated and react render the Typeahead. (4) Now you have a form value without toppingIds but the dom does have the hidden input rendered 😅

**Reply by halljus** [reply]
Hmm, in that case, I guess I fail to see how `useControl` was supposed to enable conditional rendering? It seems like we'll still need to unconditionally render the hidden input, right? (I'm fine with this, just trying to understand.)

**Reply by edmundhung** [reply]
It will be validated fine because we will always capture the latest formdata before validation. This is more a limitation of the form value with conditional rendering at the moment which I can imagine solving by adding a mechanism for useControl to notify the form context to update the form value again when register is called 🤔

**Reply by halljus** [reply]
Interesting. And I think I just ran into another edge case where this is applicable, too. 😅

**Reply by edmundhung** [reply]
Sorry for the troubles 😅 I guess you might hit case 2 a lot if you have many conditionally rendered field and relies on field.value heavily. I will look into it over the weekend. But do you have any concern with case 1 too?

**Reply by halljus** [reply]
> Sorry for the troubles 😅 I guess you might hit case 2 a lot if you have many conditionally rendered field and relies on field.value heavily. I will look into it over the weekend. But do you have any concern with case 1 too?

No apology necessary, we're incredibly grateful you're so responsive and willing to work with us on this stuff! 🙏 It feels like Conform is _so_ close to being able to handle anything!

I'm honestly not sure about case 1, but my instinct is that it's fine? I think the issue we just encountered a few minutes ago is just another instance of case 2. If I have time, I'll make another repro just in case.

Thanks for all your work!

**Reply by halljus** [reply]
Here's a reproduction of what we just encountered. More of case 2?

https://stackblitz.com/edit/remix-run-remix-5b5kux?file=app%2Froutes%2F_index.tsx

**Reply by edmundhung** [reply]
That's case 2

> The Observe `form.value` retains the default/initial `customTopping` value. 👍

It is retained only because the customTopping is still rendered when you remove the topping and Conform doesn't know the `customTopping` field is removed later. If you try to submit the form after removing the topping, you will see the `customTopping` value to be gone.

**Reply by halljus** [reply]
Right, so to accomplish the goal (keeping value/dirty states in sync), we can't conditionally render `customTopping` either. We have to fall back to rendering it unconditionally and moving the condition to the `hidden` attribute like we were doing before. Right? At least for now, until you can find a sweet fix! 😁

**Reply by halljus** [reply]
Hey Edmund, have you had a chance to consider a solution to case 2? No rush, just curious!

**Reply by edmundhung** [reply]
It will be fixed by https://github.com/edmundhung/conform/pull/491. Implementation is already done. I just need to write some tests.

**Reply by halljus** [reply]
Here's an updated reproduction for v1.1.0 (cross-posting it here since the issue where I originally provided it is now closed and it probably belongs here anyway):

https://stackblitz.com/edit/remix-run-remix-ddmnpg?file=app%2Froutes%2F_index.tsx

**Comment by bzitzow-osh** [comment]
> I think I should also start putting a PR up with all the examples updated to use the useControl hook.

@edmundhung did that set of examples ever get published?
